packages/lit-html/src/directives/async-replace.ts

```diff
@@ -10,69 +10,127 @@ import {AsyncDirective} from '../async-directive.js';
 
 type Mapper<T> = (v: T, index?: number) => unknown;
 
-class AsyncReplaceDirective extends AsyncDirective {
-  private _value?: AsyncIterable<unknown>;
-  private _reconnectResolver?: () => void;
-  private _reconnectPromise?: Promise<void>;
+// When a directive is connected and actively awaiting a result from a promise,
+// the directive will be stored in `promiseToDirectiveMap`. The resolver will
+// get a reference back to the directive to perform the commit via the WeakMap
+// rather than closing over the directive instance, which would hold the
+// directive until the promise resolved.
+const promiseToDirectiveMap: WeakMap<Promise<unknown>, AsyncReplaceDirective> =
```

Comment:
What if the same iterable and thus promise is used in multiple directives? Does this need to be (can it be?) on the directive instance?

Reply:
Man, that's a good point (more for `until`, since I think that iterables are more likely to return fresh promises for each `next()` call). I can see if the WeakMaps can live on the directive instance if we pull it off first and only close over the WeakMap. Lemme try that.

Reply:
To address this, went ahead and refactored it into a `DisconnectableAwaiter` that wraps up the logic around indirecting the `this` ref to the directive, storing a value that resolves while disconnected, and flushing it back to the `thenFn` if resolved while disconnected.

Reply:
Added tests for same promise/iterable passed to multiple directive instances as well.